packages/transaction-controller/src/TransactionController.ts

```diff
@@ -556,14 +621,117 @@ export class TransactionController extends BaseControllerV1<
 
     this.addPendingTransactionTrackerListeners();
 
+    // when transactionsController state changes
+    // check for pending transactions and start polling if there are any
+    this.subscribe(this.#checkForPendingTransactionAndStartPolling);
+
+    // TODO once v2 is merged make sure this only runs when
+    // selectedNetworkClientId changes
     onNetworkStateChange(() => {
       log('Detected network change', this.getChainId());
+      this.pendingTransactionTracker.onStateChange();
       this.onBootCleanup();
     });
 
     this.onBootCleanup();
+    this.messagingSystem.subscribe(
+      'NetworkController:stateChange',
+      (_, patches) => {
+        if (this.enableMultichain) {
+          const networkClients = this.getNetworkClientRegistry();
+          patches.forEach(({ op, path }) => {
+            if (op === 'remove' && path[0] === 'networkConfigurations') {
+              const networkClientId = path[1] as NetworkClientId;
+              delete networkClients[networkClientId];
+            }
+          });
+
+          this.#refreshTrackingMap(networkClients);
+          this.#refreshEtherscanRemoteTransactionSources(networkClients);
+        }
+      },
+    );
+    if (this.enableMultichain) {
+      this.#initTrackingMap();
+    }
   }
 
+  setEnableMultichain(enableMultichain: boolean) {
```

Comment:
In the interest of clarity, I appreciate the safety measure, but what do we gain with this other than additional complexity in the code paths and the potential for missed testing in the clients?

Would the multi-chain architecture not support a "single network" client if we just have the clients always provide a `networkClientId` when adding and start / stop polling for their selected network on initialisation and network change?

Reply:
see my comment [here](https://github.com/MetaMask/core/pull/3643#discussion_r1471335552). We are introducing the least amount of breaking changes via a feature flag. This ensures a smooth upgrade path for all the downstream teams.